### PR TITLE
795 ensure filenames are valid and test international filenames and text

### PR DESCRIPTION
**Pull request by hwindsor**
## Pull Request Overview

- Change to storing Filenames as hashes of the filename without extension
- Change multiple places where this is used in the system
- On AGS creation do the mapping from the filenames uploaded to the stored filenames.
- On mission creation (No filenames in the create message), just use the hashes
- Symlinks and image modifications generate the extension from the mimetype 
- Wrote test utils to store as this format and changed all test file writing to use them
- Increase testing of international data coming into the system
- Some test tweaks to ensure that unit tests do the same filename modification too

---
Don't assume that the file uploads changes arre correct. This is not an area I know well so there may well be misunderstandings.

### CODE REVIEW

File: app/modules/asset_groups/models.py [hunk 1]
```diff
@@ -786,7 +786,6 @@ def remove_annotation_from_encounter(self, encounter_guid, annot_guid):
                     'annotations' in encounter_metadata.keys()
                     and annot_guid in encounter_metadata['annotations']
                 ):
-                    breakpoint()
```
Comment (hwindsor) [inline app/modules/asset_groups/models.py:789]: that should never have been there, almost certainly my fault

File: app/modules/sightings/parameters.py [hunk 2]
```diff
@@ -84,16 +82,6 @@ def replace(cls, obj, field, value, state):
                     obj.add_asset(asset)
                     ret_val = True
 
-            elif field == 'newAssetGroup':
```
Comment (hwindsor) [inline app/modules/sightings/parameters.py:87]: Removed as created Asset group in non standard way and was never used

**Comment by colinwkingen** [comment]
Automatic rebase seems to have failed here for some reason.

File: integration_tests/test_sage_basic_identification.py [hunk 3]
```diff
@@ -121,3 +125,89 @@ def test_create_asset_group_identification(session, codex_url, test_root, login)
...
+                'idConfigs': [
+                    {
+                        'matchingSetDataOwners': 'all',
+                        'algorithms': ['hotspotter_nosv'],
+                    }
+                ],
+            },
+        ],
+    }
+    zebra2_guids = create_sighting(session, codex_url, test_root, filename, group_data)
```
Comment (karenc) [inline integration_tests/test_sage_basic_identification.py:170]: Is the filename returned in any APIs?  I was wondering whether the filename changed but I can't see if any asserts in this test check for the asset filename.

Comment (hwindsor) [reply integration_tests/test_sage_basic_identification.py:170]: Yes. I'm not sure that it is tested specifically in the unit tests but it is tested in the integration tests:
https://github.com/WildMeOrg/houston/blob/main/integration_tests/test_asset_group_sightings.py#L184
And the filename returned is the one the user sent, not the one that the file is stored as.

Comment (karenc) [reply integration_tests/test_sage_basic_identification.py:170]: But it looks like the integration test filename doesn't have any special characters so it's just stored without modifications.

Comment (hwindsor) [reply integration_tests/test_sage_basic_identification.py:170]: I think you are assuming that only filenames with international characters are modified. That is not the case. 
With this change ALL Filenames are stored as the hash of the filename plus extension. That was what was decided at the Knucklehead on Wednesday.

Comment (karenc) [reply integration_tests/test_sage_basic_identification.py:170]: Oh yeah, `secure_filename` is completely removed :+1:

**Review by colinwkingen — APPROVED** [review]
LGTM but I got rushed at the end and will revisit a bit tomorrow morning. Would love to see you glance at it too @naknomum .

**Comment by hwindsor** [comment]
> Automatic rebase seems to have failed here for some reason.

It did. I did it manually, something had got seriously messed up with I think a previous rebase and it had many of the commits duplicated. Now sorted.

File: app/utils.py [hunk 4]
```diff
@@ -141,3 +141,14 @@ def normalized_timezone_string(dt):
...
+
+# As some filenames are problematic, may contain special chars ";&/." etc store all filenames as a hash of the
+# original filename but maintain the extension
+def get_stored_filename(input_filename):
+    import hashlib
+    import os
+
+    file_without_extension, extension = os.path.splitext(input_filename)
+    extension = extension.lower()
+    return f'{hashlib.sha256(file_without_extension.encode()).hexdigest()}{extension}'
```
[the remark this answers is not in the document]
Comment (hwindsor) [reply app/utils.py:154]: Ooh, really good question. 
I would have thought that too but testing says it's fine:
`> /code/app/utils.py(152)get_stored_filename()
-> file_without_extension, extension = os.path.splitext(input_filename)
(Pdb) p input_filename
'zippy'
(Pdb) n
> /code/app/utils.py(153)get_stored_filename()
-> extension = extension.lower()
(Pdb) p file_without_extension
'zippy'
(Pdb) p extension
''
(Pdb) n
> /code/app/utils.py(154)get_stored_filename()
-> return f'{hashlib.sha256(file_without_extension.encode()).hexdigest()}{extension}'
(Pdb) p f'{hashlib.sha256(file_without_extension.encode()).hexdigest()}{extension}'
'ce98eb1cf7abbe7ed6a49678c73cc2bbfecbc756891e5223537c327deb7e73d3'
(Pdb) 
`

**Comment by naknomum** [comment]
general question:  is there a reason we are **keeping the extension** on the stored (hash-based) filename?   i would _strongly_ suggest we just drop it entirely.  main two reasons i would give:
1. it can be (and sometimes most definitely is) wrong - it does not match the actual content of the file
2. it is _quite often_ inconsistent:  `.jpg`, `.JPG`, `.jpeg`, etc.  (note lowercasing it does not always deal with consistency)

i _do_ however (of course?) think we should leave it as-is as part of the original filename we store in the db.   if a user wants to search on `my-image.jpeg`, that is what it should find in that field.

File: app/modules/asset_groups/models.py (hunk 1 above)
Comment (naknomum) [reply app/modules/asset_groups/models.py:789]: i had pulled this out in one my works-in-progress.  great, now i am going to get a conflict!  lol

File: app/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (naknomum) [reply app/utils.py:154]: ah, so it considers `foo.` to have an extension of empty-string but not None.  i _guess_ that makes sense?  :thinking:

**Comment by colinwkingen** [comment]
> general question: is there a reason we are **keeping the extension** on the stored (hash-based) filename? i would _strongly_ suggest we just drop it entirely. main two reasons i would give:
> 
>     1. it can be (and sometimes most definitely is) wrong - it does not match the actual content of the file
> 
>     2. it is _quite often_ inconsistent:  `.jpg`, `.JPG`, `.jpeg`, etc.  (note lowercasing it does not always deal with consistency)
> 
> 
> i _do_ however (of course?) think we should leave it as-is as part of the original filename we store in the db. if a user wants to search on `my-image.jpeg`, that is what it should find in that field.

Generally agree, but is there any actual harm in keeping it as part of the hash? We don't have a way to remedy faulty file extensions anyway.

File: tests/modules/test_ia_pipeline.py [hunk 5]
```diff
@@ -35,6 +35,11 @@ def test_ia_pipeline_sim_detect_response(
     creation_data = asset_group_utils.get_bulk_creation_data(
         test_root, request, 'african_terrestrial'
     )
+    description = 'International names Þröstur Sélène cédric characters &%$* ¼ ©,® ™ m².'
```
Comment (naknomum) [inline tests/modules/test_ia_pipeline.py:38]: very nice.  the more we test unicode _everywhere_ the happier we will be later.

**Review by naknomum — APPROVED** [review]
all looks good to me, other than my comment about dropping extensions.

also, i would love to see a test that just goes through an ever-growing list `awful_filenames = [.....]` and tests a file with each one.  but i am totally fine to do this as a separate ticket on my own.  it would be very cathartic for me.  haha

File: app/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (colinwkingen) [reply app/utils.py:154]: I think so too? \<shrug emoji\>

**Comment by colinwkingen** [comment]
> all looks good to me, other than my comment about dropping extensions.
> 
> also, i would love to see a test that just goes through an ever-growing list `awful_filenames = [.....]` and tests a file with each one. but i am totally fine to do this as a separate ticket on my own. it would be very cathartic for me. haha

Yeah and every time we find a new unhandled case we can add it to the sh*t list. Would love to see that.

**Comment by naknomum** [comment]
> > general question: is there a reason we are **keeping the extension** on the stored (hash-based) filename? i would _strongly_ suggest we just drop it entirely. main two reasons i would give:
> > ```
> > 1. it can be (and sometimes most definitely is) wrong - it does not match the actual content of the file
> > 
> > 2. it is _quite often_ inconsistent:  `.jpg`, `.JPG`, `.jpeg`, etc.  (note lowercasing it does not always deal with consistency)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > i _do_ however (of course?) think we should leave it as-is as part of the original filename we store in the db. if a user wants to search on `my-image.jpeg`, that is what it should find in that field.
> 
> Generally agree, but is there any actual harm in keeping it as part of the hash? We don't have a way to remedy faulty file extensions anyway.

the only real "harm" is it is potentially misleading noise.  if we have no _guarantee_ that a file called `XXXXXXXX.jpeg` is _actually_ a JPEG image, then why bother?

also, technically we _do_ have a way to remedy the extension -- just not at this point in the pipeline of processing, unfortunately.  (`git_store` uses `magic.from_file()` to find the _true_ content-type from the file itself.  just later on...)  this is how i have handled this in the past when we _needed_ extensions -- used file-magic to set the correct thing.  (see: apparent work, sito, etc.)   but it still begs the question: why have it at all.  :grinning:

**Comment by colinwkingen** [comment]
Lets drop the extensions in this PR, and make the unholy_filename_test in another.

**Comment by hwindsor** [comment]
> > general question: is there a reason we are **keeping the extension** on the stored (hash-based) filename? i would _strongly_ suggest we just drop it entirely. main two reasons i would give:
> > ```
> > 1. it can be (and sometimes most definitely is) wrong - it does not match the actual content of the file
> > 
> > 2. it is _quite often_ inconsistent:  `.jpg`, `.JPG`, `.jpeg`, etc.  (note lowercasing it does not always deal with consistency)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > i _do_ however (of course?) think we should leave it as-is as part of the original filename we store in the db. if a user wants to search on `my-image.jpeg`, that is what it should find in that field.
> 
> Generally agree, but is there any actual harm in keeping it as part of the hash? We don't have a way to remedy faulty file extensions anyway.

The reason for keeping the extension was that the assets class stores it separately and it is used separately. 
https://github.com/WildMeOrg/houston/blob/main/app/modules/assets/models.py#L40
https://github.com/WildMeOrg/houston/blob/main/app/modules/assets/models.py#L204-L209 
called from :
https://github.com/WildMeOrg/houston/blob/main/app/modules/assets/models.py#L219-L222
so on the principle of not changing more than needed I kept it. 
But happpy to take feedback on what should happen with this code if the extension is not stored anymore.

File: app/modules/asset_groups/models.py (hunk 1 above)
Comment (hwindsor) [reply app/modules/asset_groups/models.py:789]: saweeee

**Review by naknomum — APPROVED** [review]
new mods look good
